lib/stages/BeamBufferSort.cpp

```diff
@@ -0,0 +1,414 @@
+    //INFO("Filling empty time2 {:d}", fake_ctime.tv_sec);
+    fake_freq_bin = freq_idx;
+    fill_freq_meta((FreqIDBeamMetadata*) &sort_queue[time_idx][freq_idx][0], fake_freq_bin,
+        fake_fpga_seq_start, fake_ctime, 0, dset_id_t::null, beam_number, ra, dec, scaling);
+    memset(&sort_queue[time_idx][freq_idx][FreqIDBeamMeta_size], 0, queue_frame_size * num_pol);
+    queue_status[time_idx][freq_idx] = 1;
+	    
+}
+
+
+void BeamBufferSort::main_thread(){
+    uint32_t freq_bin;
+    int time_offset;
+    uint64_t frame_nr;
+    uint8_t* output = nullptr;
+    uint8_t* input = nullptr;
+    uint32_t advance_nframe;
+    uint32_t copy_size;
+    uint64_t fpga_seq_start;
+    uint64_t fpga_seq_start0;
+    uint64_t fpga_seq_start_at_dump; 
+    uint64_t frame0 = 0;
+    uint32_t beam_number;
+    uint32_t leftover_in_frame;
+    uint32_t empty_out_buff;
+    uint32_t dump_frame_count;
+    double ra;
+    double dec;
+    double scaling;
+    uint32_t in_frame_dump_offset = 0 ;  // Dump Offset in the incoming frame 
+    uint32_t output_buff_offset = 0; 
+    uint32_t src_start;
+    uint32_t dest_start;
+
+    // The vector for queue start metadata
+    std::vector<FreqIDBeamMetadata*> curr_queue0_metadata;
+    for (uint32_t i = 0; i < total_freq_chan; i++)
+	curr_queue0_metadata.push_back(nullptr);
+
+
+    timespec ctime0;
+    timespec ctime_at_dump;
+
+    FreqIDBeamMetadata* in_metadata = nullptr;
+    BeamMetadata* non_freq_metadata = nullptr;
+
+    frameID input_frame_id(in_buf); // Input frame id
+    std::vector<frameID> output_frame_ids; 
+    for (auto& out_buf : out_bufs){
+        output_frame_ids.push_back(frameID(out_buf));
+    }
+    
+    FreqIDBeamMetadata * temp_freq_metadata = (FreqIDBeamMetadata*)malloc(FreqIDBeamMeta_size);
+
+    while (!stop_thread) {
+	// Receive a new frame
+        input = wait_for_full_frame(in_buf, unique_name.c_str(), input_frame_id);
+	if (input == nullptr)
+            break;
+        
+	// Get the metadata for each frame.
+        if (has_freq_bin) {
+            // Frequency ID is in the metdata.
+            in_metadata = (FreqIDBeamMetadata*)get_metadata(in_buf, input_frame_id);
+            freq_bin = in_metadata -> frequency_bin;
+            fpga_seq_start = in_metadata -> fpga_seq_start;
+        }
+        else {
+	    // Frequency ID is not in the metadata. will add it to the metadata.
+            non_freq_metadata = (BeamMetadata*)get_metadata(in_buf, input_frame_id);
+            fpga_seq_start = non_freq_metadata -> fpga_seq_start;
+            const uint32_t num_freq_per_stream = Telescope::instance().num_freq_per_stream();
+            for (uint32_t f = 0; f < num_freq_per_stream; ++f) {
+                freq_bin = Telescope::instance().to_freq_id(non_freq_metadata -> stream_id, f);
+            }
+	    // Build the full Frequency IDed metadata
+	    in_metadata = temp_freq_metadata;
+	    nonfreq_meta_2_freq_meta(non_freq_metadata, in_metadata, freq_bin);
+        }
```

Comment:
So sometimes the incoming beam metadata doesn't contain a frequency bin?

Reply:
It needs a frequency mapping, but some other stage may add a frequency bin to it. For example the mergebeambuffer stage. So it would be better to have a check.